packages/statemanager/src/statefulBinaryTreeStateManager.ts

```diff
@@ -0,0 +1,802 @@
+  encodeBinaryTreeLeafBasicData,
+  equalsBytes,
+  generateBinaryTreeChunkSuffixes,
+  generateBinaryTreeCodeStems,
+  getBinaryTreeKeyForStorageSlot,
+  getBinaryTreeStem,
+  hexToBigInt,
+  hexToBytes,
+  padToEven,
+  setLengthLeft,
+  setLengthRight,
+  short,
+  unprefixedHexToBytes,
+} from '@ethereumjs/util'
+import { blake3 } from '@noble/hashes/blake3'
+import debugDefault from 'debug'
+import { keccak256 } from 'ethereum-cryptography/keccak.js'
+
+import { OriginalStorageCache } from './cache/originalStorageCache.js'
+import { modifyAccountFields } from './util.js'
+
+import type { Caches } from './cache/caches.js'
+import type { BinaryTreeState, StatefulBinaryTreeStateManagerOpts } from './types.js'
+import type {
+  AccountFields,
+  BinaryTreeAccessWitnessInterface,
+  BinaryTreeAccessedStateWithAddress,
+  StateManagerInterface,
+  StorageDump,
+  StorageRange,
+} from '@ethereumjs/common'
+import type {
+  BinaryTreeExecutionWitness,
+  GenesisState,
+  PrefixedHexString,
+  StoragePair,
+} from '@ethereumjs/util'
+import type { Debugger } from 'debug'
+
+const ZEROVALUE = '0x0000000000000000000000000000000000000000000000000000000000000000'
+export class StatefulBinaryTreeStateManager implements StateManagerInterface {
+  protected _debug: Debugger
+  protected _caches?: Caches
+
+  preStateRoot: Uint8Array
+  originalStorageCache: OriginalStorageCache
+  hashFunction: (input: Uint8Array) => Uint8Array
+
+  protected _tree: BinaryTree
+
+  protected _checkpointCount: number
+
+  // Post-state provided from the executionWitness.
+  // Should not update. Used for comparing our computed post-state with the canonical one.
+  private _postState: BinaryTreeState = {}
+
+  /**
+   * StateManager is run in DEBUG mode (default: false)
+   * Taken from DEBUG environment variable
+   *
+   * Safeguards on debug() calls are added for
+   * performance reasons to avoid string literal evaluation
+   * @hidden
+   */
+  protected readonly DEBUG: boolean = false
+
+  private keccakFunction: Function
+
+  constructor(opts: StatefulBinaryTreeStateManagerOpts) {
+    // Skip DEBUG calls unless 'ethjs' included in environmental DEBUG variables
+    // Additional window check is to prevent vite browser bundling (and potentially other) to break
+    this.DEBUG =
+      typeof window === 'undefined' ? (process?.env?.DEBUG?.includes('ethjs') ?? false) : false
+
+    this._checkpointCount = 0
+
+    // TODO: Do we need to check this?
+    if (opts.common?.isActivatedEIP(7864) === false) {
+      throw EthereumJSErrorWithoutCode('EIP-7864 required for binary tree state management')
+    }
+
+    this.hashFunction = opts.hashFunction ?? blake3
+    this._tree =
+      opts.tree ??
+      new BinaryTree({
+        hashFunction: this.hashFunction,
+        db: new MapDB<string, string | Uint8Array>(),
+        useRootPersistence: false,
+        cacheSize: 0,
+      })
+    this._debug = debugDefault('statemanager:binarytree:stateful')
```

Comment:
I think we might be able to skip the `stateful` for now.  That distinction doesn't make sense in Binary Trees since we aren't building a separate equivalent.

Reply:
Removed. I guess we might still want a stateless implementation at some point, but clearly not a priority right now. I think I'd still keep the Stateful naming in the class name though